File: benchmarks/go/b7_concurrent.go
package main
import ("fmt"; "sync"; "sync/atomic")
func main() {
    var counter int64
    var wg sync.WaitGroup
    for i := 0; i < 8; i++ {
        wg.Add(1)
        go func() {
            defer wg.Done()
            for j := 0; j < 10000000; j++ { atomic.AddInt64(&counter, 1) }
        }()
    }
    wg.Wait()
    fmt.Println(counter)
}

File: benchmarks/julia/b7_concurrent.jl
using Base.Threads

const OPS = 10_000_000

# Must be launched with `julia -t 8` — harness does this for b7 specifically.
# Otherwise nthreads()=1 and @threads runs serially, defeating the contention
# scenario this benchmark exists to measure.
mutable struct Counter
    @atomic value::Int64
end

function main()
    @assert nthreads() == 8 "expected 8 threads, got $(nthreads()) — invoke with: julia -t 8"
    c = Counter(0)
    Threads.@threads for _ in 1:8*OPS
        @atomic :monotonic c.value += 1
    end
    println(c.value)
end

main()
